arduino/opencr_arduino/opencr/libraries/turtlebot3/turtlebot3_waffle/turtlebot3_core/turtlebot3_core.ino

```diff
@@ -128,6 +130,51 @@ static uint8_t battery_state   = BATTERY_POWER_OFF;
 
 
 
+/*******************************************************************************
+* Time Interpolation function
+*******************************************************************************/
+ros::Time add_micros(ros::Time & t, uint32_t _micros)
+{
+  uint32_t sec, nsec;
+
+  /*
+   * the interpolation floors the current time
+   * stamp for integrity of precision.
+   */
+  sec  = _micros / 1000000 + t.sec;
+  nsec = _micros % 1000000 + 1000 * (t.nsec / 1000);
+  if (nsec >= 1e9) {
+    sec++, nsec--;
+  }
+  return ros::Time(sec, nsec);
+}
+
+/*******************************************************************************
+* Update the base time for interpolation
+*******************************************************************************/
+void update_time()
+{
+  uint32_t n_micros = micros();
+  ros::Time next = nh.now();
+  if (next.sec < current_time.sec) {
+    /*
+     * bad time was received, so we
```

Comment:
> simulated time can go backwards, which this does not seem robust to.

That is a really good point. I'll see if I can probe `use_sim_time` in `rosserial`.

This PR also needs to make the changes for the Burger, I just haven't tested it on that platform yet.